controller/node_controller.go

```diff
@@ -1021,3 +871,215 @@ func BackingImageDiskFileCleanup(node *longhorn.Node, bi *longhorn.BackingImage,
 		delete(bi.Spec.Disks, diskUUID)
 	}
 }
+
+func (nc *NodeController) checkMonitor(node *longhorn.Node) (monitor.Monitor, error) {
+	if node == nil {
+		return nil, nil
+	}
+
+	if nc.monitor != nil {
+		return nc.monitor, nil
+	}
+
+	monitor, err := monitor.NewNodeMonitor(nc.logger, nc.eventRecorder, nc.ds, node, nc.enqueueNodeForMonitor)
+	if err != nil {
+		return nil, err
+	}
+
+	nc.monitor = monitor
+
+	return monitor, nil
+}
+
+func (nc *NodeController) enqueueNodeForMonitor(key string) {
+	nc.queue.Add(key)
+}
+
+func (nc *NodeController) syncOrphans(node *longhorn.Node, onDiskReplicaDirectoryNames map[string]map[string]string) error {
+	if err := nc.cleanupOrphans(node, onDiskReplicaDirectoryNames); err != nil {
+		return errors.Wrapf(err, "failed to clean up orphans")
+	}
+
+	if err := nc.createOrphans(node, onDiskReplicaDirectoryNames); err != nil {
+		return errors.Wrapf(err, "failed to create orphans")
+	}
+
+	if autoDeletion, err := nc.ds.GetSettingAsBool(types.SettingNameOrphanAutoDeletion); err == nil {
+		if autoDeletion {
+			if err := nc.ds.DeleteAllOrphansForNode(node.Name); err != nil {
+				return errors.Wrapf(err, "failed to delete orphans for node %v", node.Name)
+			}
+		}
+	} else {
+		return errors.Wrapf(err, "failed to get %v setting", types.SettingNameOrphanAutoDeletion)
+	}
+
+	return nil
+}
+
+func (nc *NodeController) cleanupOrphans(node *longhorn.Node, onDiskReplicaDirectoryNames map[string]map[string]string) error {
+	labelSelector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			types.GetLonghornLabelKey(types.LonghornLabelNode): nc.controllerID,
+		},
+	})
+	if err != nil {
+		return err
+	}
+	orphans, err := nc.ds.ListOrphansBySelectorRO(labelSelector)
+	if err != nil {
+		return errors.Wrapf(err, "unable to list orphans with label %v=%v",
+			types.GetLonghornLabelKey(types.LonghornLabelNode), nc.controllerID)
+	}
+
+	for _, orphan := range orphans {
+		if orphan.Status.OwnerID != nc.controllerID {
+			continue
+		}
+
+		if orphan.Spec.NodeID == orphan.Status.OwnerID {
```

Comment:
Do we need to check these node ID fields here? The orphan controller should handle it and tell the orphan object final state.